### ISSUE TITLE
job terminated before all instances finished

### ISSUE THREAD

**Issue by nclowell** [RobertsLab/resources#858]
I'm running model on mox, which is a python script. The model is slow so I use a bash script to run simulation replicates on multiple instances, and then can combine the output later. I think this means it's running on multiple cores, because it takes about the same time for many instances to complete as it does one instance, etc. Anyway, when I do this on my own computer, it works just fine. I think it's mostly working just fine on mox too, except on one job it ran almost all instances to completion before the job ended without errors. Another time only one (of 28) instance had gone to completion before the job ended without errors. In all 3 jobs, the first instance never completed. Just wondering if there's something I don't understand about.... computers or maybe mox specifically as to why this is happening! Any idea how I can make sure all instances complete before the job terminates?

Just in case it's helpful, here's the text of the bash script I use to call the model onto different instances:

```
#!/bin/bash

#########################################################################
# Call GRs model to run on X cores with Y reps per core
# example usage:
# sh run_GRs_xcores.sh GRs_v2.21.py 001 3 1

# args from command line
model_file=$1 # path to model script
batch_ID=$2 # batch ID, unique to parameter set
n_cores=$3 # number of cores to spread simulations onto
n_reps_per_core=$4 # number of reps to run per core / process

for srep in $(seq 1 $(($n_cores-1))); do
    python $model_file $batch_ID $n_reps_per_core $srep &
done

python $model_file $batch_ID $n_reps_per_core $n_cores

echo '\n---Finished running simulations---'
        
```

**Comment by kubu4** [comment]
We please need to see the actual SBATCH script used to run on Mox. Also, it would be helpful to see the end of the SLURM output file (run: `tail -n 50 <slurm.out>` to just capture the last 50 lines - this is usually sufficient info).

**Comment by nclowell** [comment]
Sorry, my bad! 

Here's the text of the sbatch script:

```
#!/bin/bash
## Job Name
#SBATCH --job-name=nl_testmodel_02
## Allocation Definition
#SBATCH --account=srlab
#SBATCH --partition=srlab
## Resources
## Nodes (We only get 1, so this is fixed)
#SBATCH --nodes=1
## Walltime (days-hours:minutes:seconds format)
#SBATCH --time=3-00:00:00
## Memory per node
#SBATCH --mem=475G
#SBATCH --mail-type=ALL
#SBATCH --mail-user=nclowell@uw.edu
## Specify the working directory for this job
#SBATCH --chdir=/gscratch/scrubbed/nclowell/nl_testmodel_02
# nclowell 20200302
# The purpose of this script is to run my genetic risks model on mox, mostly as a test of how
# long it takes and to make sure I can do it; second try, because overestimated memory improvement on hyak


sh /gscratch/scrubbed/nclowell/nl_testmodel_02/run_GRs_xcores.sh /gscratch/scrubbed/nclowell/nl_testmodel_02/GRs_v2.28.py 002 28 1

```

And here's the last 50 lines of the slurm.out

```
1450 2020-03-04 14:42:38
1451 2020-03-04 14:44:27
1452 2020-03-04 14:46:16
1453 2020-03-04 14:48:05
1454 2020-03-04 14:49:56
1455 2020-03-04 14:51:47
1456 2020-03-04 14:53:38
1457 2020-03-04 14:55:32
1458 2020-03-04 14:57:23
1459 2020-03-04 14:59:12
1460 2020-03-04 15:01:02
1461 2020-03-04 15:02:51
1462 2020-03-04 15:04:39
1463 2020-03-04 15:06:28
1464 2020-03-04 15:08:17
1465 2020-03-04 15:10:06
1466 2020-03-04 15:11:55
1467 2020-03-04 15:13:45
1468 2020-03-04 15:15:35
1469 2020-03-04 15:17:24
1470 2020-03-04 15:19:16
1471 2020-03-04 15:21:06
1472 2020-03-04 15:22:56
1473 2020-03-04 15:24:46
1474 2020-03-04 15:26:36
1475 2020-03-04 15:28:28
1476 2020-03-04 15:30:20
1477 2020-03-04 15:32:11
1478 2020-03-04 15:34:02
1479 2020-03-04 15:35:54
1480 2020-03-04 15:37:46
1481 2020-03-04 15:39:39
1482 2020-03-04 15:41:33
1483 2020-03-04 15:43:28
1484 2020-03-04 15:45:21
1485 2020-03-04 15:47:15
1486 2020-03-04 15:49:09
1487 2020-03-04 15:51:03
1488 2020-03-04 15:52:56
1489 2020-03-04 15:54:49
1490 2020-03-04 15:56:41
1491 2020-03-04 15:58:34
1492 2020-03-04 16:00:27
1493 2020-03-04 16:02:19
1494 2020-03-04 16:04:11
1495 2020-03-04 16:06:02
1496 2020-03-04 16:08:02
1497 2020-03-04 16:09:54
1498 2020-03-04 16:12:08
1499 2020-03-04 16:14:14
\n---Finished running simulations---

```
All of the slurm.out is error free, and it's just spitting out a timestamp that the script is supposed to every time step.

One other difference that I just noticed now, when comparing running this on my computer and on the super computer, is that usually each instance of the model is running roughly concurrently, so I get multiple of the same numbered time step printing to the screen at the same time, like:

1492 2020-03-04 16:00:27
1492 2020-03-04 16:02:19
1492 2020-03-04 16:04:11
1493 2020-03-04 16:06:02
1493 2020-03-04 16:08:02
1493 2020-03-04 16:09:54
1494 2020-03-04 16:12:08

But looking through the slurm.out everything is neatly ordered and replicates are appearing in sequence as opposed to concurrently.

**Comment by kubu4** [comment]
Let' walk through this script first. This will help me make sure I'm following and thinking about this correctly. Hopefully it will help you help me. :)

```shell
#!/bin/bash

#########################################################################
# Call GRs model to run on X cores with Y reps per core
# example usage:
# sh run_GRs_xcores.sh GRs_v2.21.py 001 3 1

# args from command line
model_file=$1 # path to model script
batch_ID=$2 # batch ID, unique to parameter set
n_cores=$3 # number of cores to spread simulations onto
n_reps_per_core=$4 # number of reps to run per core / process

for srep in $(seq 1 $(($n_cores-1))); do
    python $model_file $batch_ID $n_reps_per_core $srep &
done

python $model_file $batch_ID $n_reps_per_core $n_cores
```

---

1. You define the arguments that the script will accept:

```shell
# args from command line
model_file=$1 # path to model script
batch_ID=$2 # batch ID, unique to parameter set
n_cores=$3 # number of cores to spread simulations onto
n_reps_per_core=$4 # number of reps to run per core / process
```

2. Loop to process each batch ID

- `for srep in $(seq 1 $(($n_cores-1))); do`: This uses the number of cores minus one as the end value to the `seq` command (i.e. ends up being `seq 1 27`) and stores the current value in `srep`.

- `python $model_file $batch_ID $n_reps_per_core $srep &`: This calls `python` and the `$model_file`. It provides the `$model_file` with the following arguments. The arguments are taken from your SBATCH script `sh /gscratch/scrubbed/nclowell/nl_testmodel_02/run_GRs_xcores.sh /gscratch/scrubbed/nclowell/nl_testmodel_02/GRs_v2.28.py 002 28 1`

  - `$batch_ID1`: `002`
  - `$n_reps_per_core`: `1`
  - `$srep`: `1..27`
  - `&`: This runs the script "asynchronously" which allows multiple instances of the Python script to run without having to wait for the previous one to finish by running each instance on a different CPU core. Very cool!

3. Runs Python `$model_file` one last time (not sure why; could just change the for loop to end with `$n_cores`):

- `python $model_file $batch_ID $n_reps_per_core $n_cores`: Basically runs the model one more time using `$n_cores` (28) instead of `$srep` value in previous for loop.

---

So, that all looks good to me.

However, looking at the job from the SBATCH script you posted above, I'm not entirely sure what's going on.

When I look at the contents of the specified working directory (`/gscratch/scrubbed/nclowell/nl_testmodel_02`), I see this:

```shell
drwxr-xr-x 2 nclowell hyak-srlab   4096 Mar  3 02:55 batch_002_srep_10_outfiles
drwxr-xr-x 2 nclowell hyak-srlab   4096 Mar  4 14:54 batch_002_srep_11_outfiles
drwxr-xr-x 2 nclowell hyak-srlab   4096 Mar  4 15:54 batch_002_srep_12_outfiles
drwxr-xr-x 2 nclowell hyak-srlab   4096 Mar  3 02:53 batch_002_srep_13_outfiles
drwxr-xr-x 2 nclowell hyak-srlab   4096 Mar  4 15:40 batch_002_srep_14_outfiles
drwxr-xr-x 2 nclowell hyak-srlab   4096 Mar  4 15:19 batch_002_srep_15_outfiles
drwxr-xr-x 2 nclowell hyak-srlab   4096 Mar  4 15:58 batch_002_srep_16_outfiles
drwxr-xr-x 2 nclowell hyak-srlab   4096 Mar  4 15:11 batch_002_srep_17_outfiles
drwxr-xr-x 2 nclowell hyak-srlab   4096 Mar  4 14:50 batch_002_srep_18_outfiles
drwxr-xr-x 2 nclowell hyak-srlab   4096 Mar  4 15:11 batch_002_srep_19_outfiles
drwxr-xr-x 2 nclowell hyak-srlab   4096 Mar  4 15:56 batch_002_srep_1_outfiles
drwxr-xr-x 2 nclowell hyak-srlab   4096 Mar  4 15:06 batch_002_srep_20_outfiles
drwxr-xr-x 2 nclowell hyak-srlab   4096 Mar  3 02:19 batch_002_srep_21_outfiles
drwxr-xr-x 2 nclowell hyak-srlab   4096 Mar  4 15:07 batch_002_srep_22_outfiles
drwxr-xr-x 2 nclowell hyak-srlab   4096 Mar  4 15:37 batch_002_srep_23_outfiles
drwxr-xr-x 2 nclowell hyak-srlab   4096 Mar  4 14:39 batch_002_srep_24_outfiles
drwxr-xr-x 2 nclowell hyak-srlab   4096 Mar  3 02:24 batch_002_srep_25_outfiles
drwxr-xr-x 2 nclowell hyak-srlab   4096 Mar  4 16:05 batch_002_srep_26_outfiles
drwxr-xr-x 2 nclowell hyak-srlab   4096 Mar  4 14:41 batch_002_srep_27_outfiles
drwxr-xr-x 2 nclowell hyak-srlab   4096 Mar  4 16:16 batch_002_srep_28_outfiles
drwxr-xr-x 2 nclowell hyak-srlab   4096 Mar  3 02:41 batch_002_srep_2_outfiles
drwxr-xr-x 2 nclowell hyak-srlab   4096 Mar  4 15:04 batch_002_srep_3_outfiles
drwxr-xr-x 2 nclowell hyak-srlab   4096 Mar  4 14:58 batch_002_srep_4_outfiles
drwxr-xr-x 2 nclowell hyak-srlab   4096 Mar  4 15:13 batch_002_srep_5_outfiles
drwxr-xr-x 2 nclowell hyak-srlab   4096 Mar  4 15:37 batch_002_srep_6_outfiles
drwxr-xr-x 2 nclowell hyak-srlab   4096 Mar  4 15:17 batch_002_srep_7_outfiles
drwxr-xr-x 2 nclowell hyak-srlab   4096 Mar  4 14:15 batch_002_srep_8_outfiles
drwxr-xr-x 2 nclowell hyak-srlab   4096 Mar  3 02:45 batch_002_srep_9_outfiles
-rwxrwxrwx 1 nclowell hyak-srlab  72362 Feb 27 16:14 GRs_v2.28.py
-rwxrwxrwx 1 nclowell hyak-srlab    841 Mar  2 10:09 nl_testmodel_02.sh
-rwxrwxrwx 1 nclowell hyak-srlab    650 Feb 27 16:06 run_GRs_xcores.sh
-rw-r--r-- 1 nclowell hyak-srlab 805076 Mar  4 16:16 slurm-2005310.out
```

To me, it looks like everything worked (i.e. the for loop ran your script 28 times). However, when searching the `slurm-2005310.out` file for each batch, this is what I see:

```shell
[samwhite@mox1 nl_testmodel_02]$ grep batch slurm-2005310.out | sort --version-sort
batch_002_srep_1
batch_002_srep_3
batch_002_srep_4
batch_002_srep_5
batch_002_srep_6
batch_002_srep_7
batch_002_srep_8
batch_002_srep_11
batch_002_srep_12
batch_002_srep_14
batch_002_srep_15
batch_002_srep_16
batch_002_srep_17
batch_002_srep_18
batch_002_srep_19
batch_002_srep_20
batch_002_srep_22
batch_002_srep_23
batch_002_srep_24
batch_002_srep_26
batch_002_srep_27
batch_002_srep_28
```

There were definitely "missed" `sreps`.

So, why are there output folders for each? 

All this is really beside the point, though.

If this works on your computer, why bother with Mox? Does this take a ridiculously long time on your own computer?

**Comment by nclowell** [comment]
It does take a ridiculously long time on my computer! And my computer is fine for all sorts of testing - I can pick parameter values that make the model run quicker for debugging (smaller simulated population sizes over shorter time steps), but the output isn't meaningful biologically. When I have to run the sensitivity analysis / final simulations, I'm definitely going to need a more powerful machine.

Thanks for walking through the script with me! Your notes all seem correct to my best understanding. When I first researched how to call multiple instances at once, whatever site I was on had a recommendation for calling the last instance not in the loop. I honestly don't remember now why though.

At some point in time I thought it made since to make folders for each set of replicates, but recently I realized I named everything wisely enough that they should be able to go into a single folder. It's on my list of fixes.

Let me know what else I can share to help figure out why some sreps didn't run to completion / got skipped! Thank you

**Comment by kubu4** [comment]
> At some point in time I thought it made since to make folders for each set of replicates

So, does this mean you created all of the output folders and they are _not_ created by the script?

**Comment by kubu4** [comment]
I just thought of something. How do you know the SLURM job finished successfully? I looked at the exit code of the job of the SBATCH script listed above and do see that it has an exit code of zero (which means no errors), but it just dawned on me that the use of the `&` at the end of your command is the likely culprit in things not all getting processed. I'll explain after I hear back from you.

**Comment by nclowell** [comment]
Nope, the script makes the directories.

I guess I don't know that it was successful, or what exactly successful means. I meant I didn't get any errors and I got the notice that jobs end with exit code 0.

**Comment by kubu4** [comment]
Gotcha. Thanks.

Here's my thinking on what's happening. When you use the `&`, it creates a "subshell" to run the script. I think that the for loop completes, but some instances of the `$model_file` are still running in a subshell that SLURM can't "see". So, after the for loop finishes, SLURM sees things as complete, but isn't tracking the process of the other subshells, and exits without error.

This would potentially explain why you get results for most of them, but not all; different reps take longer/shorter than others and complete at different times.

The way to test this is to run an SBATCH with the `&` at the end of your `python` command and see what happens.

**Comment by nclowell** [comment]
I'll try that next, and get back to you, thanks!

I guess the issue is I do want it to run things simultaneously, even if the right way to do that is different that how I'm trying to now. I suspect if I take the & out it will run them sequentially?

**Comment by kubu4** [comment]
Yes, they will run sequentiality.

**Comment by kubu4** [comment]
> I guess the issue is I do want it to run things simultaneously

Don't we all! :cat:

**Comment by nclowell** [comment]
Ok! Removed the ``&`` in the bash script as you suggested, to see if the model replicates would run sequentially. And it looks like the job terminated without error but only ran the first replicate of the model. Not what I expected? Let me know if this gives you any insights about what is happening on mox.

Here's the content of the sbatch script:

```
#!/bin/bash
## Job Name
#SBATCH --job-name=nl_testpll_01
## Allocation Definition
#SBATCH --account=srlab
#SBATCH --partition=srlab
## Resources
## Nodes (We only get 1, so this is fixed)
#SBATCH --nodes=1
## Walltime (days-hours:minutes:seconds format)
#SBATCH --time=3-00:00:00
## Memory per node
#SBATCH --mem=475G
#SBATCH --mail-type=ALL
#SBATCH --mail-user=nclowell@uw.edu
## Specify the working directory for this job
#SBATCH --chdir=/gscratch/scrubbed/nclowell/nl_testpll_01
# nclowell 20200317
# Run genetic risks model without & in bash script, see if runs sequentially succfessfully

sh /gscratch/scrubbed/nclowell/GR001/run_GRs_xcores.sh /gscratch/scrubbed/nclowell/GR001/GRs_v2.28.py GR001 5 1

```

And here's the bash script I used to call the model, without the ``&``

```
#!/bin/bash

#########################################################################
# Call GRs model to run on X cores with Y reps per core
# example usage:
# sh run_GRs_xcores.sh GRs_v2.21.py 001 3 1

# args from command line
model_file=$1 # path to model script
batch_ID=$2 # batch ID, unique to parameter set
n_cores=$3 # number of cores to spread simulations onto
n_reps_per_core=$4 # number of reps to run per core / process


# testing with Sam White, removing & to run sequentially and make sure job completes 
# as expected 

for srep in $(seq 1 $(($n_cores-1))); do
    python $model_file $batch_ID $n_reps_per_core $srep
done

python $model_file $batch_ID $n_reps_per_core $n_cores

echo '\n---Finished running simulations---'
```

**Comment by kubu4** [comment]
@nclowell I was looking at this and wanted to review the Python script `/gscratch/scrubbed/nclowell/GR001/GRs_v2.28.py`. However, it's not there anymore.

Could you upload that script (and, maybe re-run your job)? Also, if you're just going to do a "quick" test run, your job may begin in a more timely fashion if you request less memory (i.e. < 128GB).

**Comment by nclowell** [comment]
Thanks for the feedback! I'll run it again and will reduce the memory ask.

I'll get in touch when the new run completes. I suspect it will be the same, where the first instance completes and the rest do not. However I thought I had commented out a section of code, and hadn't, which may result in the throwing of an error. So I double-checked this one is commented out and it's running now.

Thanks again!

**Comment by nclowell** [comment]
The job completed with exit code 0, and as expected it did run them sequentially and successfully. I flubbed the sbatch script, and put the wrong directory for output files. Output files are actually in ``/gscratch/scrubbed/nclowell/nl_testpll_01``, and they're the folders named ``batch_nl_testpll_02*``. And the input files are in /gscratch/scrubbed/nclowell/nl_testpll_02, including the one you asked for, ``GRs_v2.28.py``.

In the small chance that it was the section of code (that I thought I commented out but hadn't) causing the job to finish before all instances completed, I'm going to rerun the same job and stick the ``&`` back in. I'll let you know how that goes!

Thank you!

**Comment by nclowell** [comment]
Alright! Ran the job with the ``&`` back in there, diverting new instances to the background. I also made sure the block of code was commented out that should be. And indeed, only 1 of 5 instances completed by the time the job ended with exit code 0. This confirms that it is likely the ``&`` (and not the code that previously wasn't commented out) that could have thrown a wrench in the job. Output files are all in ``/gscratch/scrubbed/nclowell/nl_testpll_03``

I guess my question is - do you know what's different about jobs on mox (or through slurm) versus a "regular" computer that the job ends before all the instances complete? Hoping to get some insights and then see if I can figure out another way to do this.

**Comment by kubu4** [comment]
> do you know what's different about jobs on mox (or through slurm) versus a "regular" computer that the job ends before all the instances complete

From my earlier post:

> Here's my thinking on what's happening. When you use the &, it creates a "subshell" to run the script. I think that the for loop completes, but some instances of the $model_file are still running in a subshell that SLURM can't "see". So, after the for loop finishes, SLURM sees things as complete, but isn't tracking the process of the other subshells, and exits without error.

**Comment by kubu4** [comment]
I'll drop a line to UW IT, though and see what they say.

**Comment by nclowell** [comment]
Sorry, I totally read that before and somehow forgot. Pandemic brain. Thanks for dropping a line at UW IT for me!

**Comment by kubu4** [comment]
UW-IT wrote back. Their response is pasted below.

TL;DR - They confirmed my suspicions about background jobs being killed after SLURM script completes. 

Suggestion is to try GNU Parallel.

However, another thought that _literally_ just popped into my head would be to run the job using an interactive node. Going this route, you wouldn't run your job using a SLURM job. You'd run it as though you were on a "normal" computer. If going this route, you'd definitely want to use the `screen` program, as an interactive node is run via SSH. If you were to stop your connection to Mox during an interactive node, it will kill any jobs you initiate. Using `screen` allows you to terminate your SSH connection without killing your job on Mox, and allows you to come back later.

The biggest downside to the above technique is that you lose the automatic logging of stdout/stderr to a log file (i.e. SLURM output file). Additionally, you will _not_ receive a notification when the job completes _and_ the amount of time you request for an interactive node will run until the time limit ends (or, when you type `exit` at the terminal prompt; it will _not_ end when your job completes). 

For example, if you were to request 12 days of wall time for your interactive node and your job finished in two days, you'd still be "using" that node for the full 12 days, even though your job ended. You'd have to manually exit the interactive node so that others could use it.

---

UW-IT repsonse:


> Hi Sam,
>
>Without getting too into the weeds, backgrounding the process with '&' does cause the the command in question to run in the background, which allows the calling script to continue.  Backgrounded processes are still, however, sub-processes of the shell that called it.  In the case of a Slurm job, this means the script will continue and will go on to terminate when it reaches the end of the script (or anything else that would cause it to terminate, like an error).  When this happens Slurm will consider the job complete and will execute its cleanup tasks, which will cause those backgrounded commands to be killed if they are still running.  The script has already completed and provided its return code at this point, so this does mean that the slurm job will show successful completion even though the backgrounded stuff may not have completed.
>
>One partial solution is to use the shell 'wait' command at some point after all of the background processes are started.  When called without any arguments it will sit there and return only when all of the backgrounded processes have completed (you can also pass a pid, in which case it will wait only for that specific process to finish).  That way all of the backgrounded processes will complete before the script finishes.  Capturing the exit code of the backgrounded processes gets much more tricky, so if knowing if the processes finish successfully is needed, it is probably easiest to call whatever you are running from as script that outputs the return code (or other indication of completion status) to a file that can then be checked before termination of the main script, which then explicitly returns an appropriate exit code based on the results of all of the backgrounded processes.
>
>Depending on the details, employing GNU Parallel is likely a good solution for these scenarios instead of manually backgrounding processes (and there is a module and some documentation, see https://wiki.cac.washington.edu/display/hyakusers/Hyak+gnu+parallel ). 
>
>All of the above said, if the scenario described is common for your group, long term you might be better off changing your partition to be non-exclusive  (like the ckpt or build partitions are now).  In other words, jobs are allocated the CPU resources they request, not always the full node.  This requires more careful and explicit resource requests (especially for CPUs) -- but in such case the scenario you are describing is simply a matter of scheduling small jobs via Slurm instead of backgrounding tasks inside of a big job (employing job dependencies or a small job-that-launches-jobs-and-monitors-them as appropriate).  It'd be a significant workflow change, but lots of "smaller" jobs are easier for the scheduler to efficiently schedule (effectively it is working at 'higher resolution').  If it influences such a decision: new partitions are being deployed this way by default now, and the 'next' cluster will likely not have exclusive partitions at all.
>
>Hope that helps.
>
>-Matt

**Comment by nclowell** [comment]
Thank you so much for relaying that information from UW-IT. I'll try GNU parallel. If I can't get that working, I'll try what you suggested re: the interactive node, using ``screen``, carefully monitoring time usage, etc. 

Thank you!!

**Comment by kubu4** [comment]
GNU Parallel has serious potential! It's one of those things that's been on my "to-do list" for a _long_ time, but it will require a bit of time to really grasp it. However, the developer of GNU Parallel actually has a walk-through tutorial to teach you how to use it. It just requires the time to sit down and do it. However, it is potentially very powerful once you have a grasp of how to use it!

**Comment by nclowell** [comment]
Cool! Well if I successfully wrap my head around it, happy to chat about what I learn with you!

**Comment by nclowell** [comment]
Just wanted to say that the hyak gnu parallel instructions were enough to get it working with my model, which is dope! In the chance that you haven't used it on mox, slurm out has the output organized by processes, which is better for me and counter intuitive to me. I expected it to capture output in real time, which would be mixing the output from all processes together as they were printed. Anyway I'm sure there's tons more about gnu parallel for me to learn but stoked that it's working for my (minimal) needs. Thanks again for getting help from UW-IT!

**Comment by kubu4** [comment]
@nclowell Can you please share how you used GNU Parallel in this instance?

**Comment by nclowell** [comment]
Of course! 

Here's the sbatch script. The points of interest are the ``--ntasks-per-node`` addition to the ``#SBATCH`` line (not sure if this is necessary but that's what they recommended), activating GNU Parallel, and then calling GNU Parallel on a script that includes all the processes you want to parallelize, one per line.

```
#!/bin/bash
## Job Name
#SBATCH --job-name=nl_test_gnupll_02
## Allocation Definition
#SBATCH --account=srlab
#SBATCH --partition=srlab
## Resources
## Nodes (We only get 1, so this is fixed)
#SBATCH --nodes=1 --ntasks-per-node=8
## Walltime (days-hours:minutes:seconds format)
#SBATCH --time=1-00:00:00
## Memory per node
#SBATCH --mem=100G
#SBATCH --mail-type=ALL
#SBATCH --mail-user=nclowell@uw.edu
## Specify the working directory for this job
#SBATCH --chdir=/gscratch/scrubbed/nclowell/nl_test_gnupll02
# nclowell 20200413
# Run genetic risks model with & gnu parallel, second try with more jobs to compare length of 
# job and make sure working as expected

# access GNU parallel
module load parallel-20170722

# specify cpu count
cpu_count=8

cat commands_for_parallel.sh | parallel -j $cpu_count
```

Here's the contents of the ``commands_for_parallel.sh`` file, for reference:

```
#!/bin/bash

# call model reps
python GRs_v2.28.py testgnu01 1 1
python GRs_v2.28.py testgnu01 1 2
python GRs_v2.28.py testgnu01 1 3
python GRs_v2.28.py testgnu01 1 4
python GRs_v2.28.py testgnu01 1 5
python GRs_v2.28.py testgnu01 1 6
python GRs_v2.28.py testgnu01 1 7
python GRs_v2.28.py testgnu01 1 8

```

Here's the first bit of the slurm out, showing you how it ordered output by processes, and not in real time. My model prints a time step for every "year," with the year number first. As all processes of the model progress concurrently here (you can see matching-ish time stamps in the two model processes below), I expected to see multiple year Xs back to back, one for each process (as it does when I ran multiple processes using the ``&`` to send processes to the background). But instead it printed the time stamps for each process, back to back.

```
Reading in parameter values
batch_testgnu01_srep_7
Initializing response variable storage objecs
Rep 0 2020-04-13 17:33:28
Initializing starting population
Pre-farm period
0 2020-04-13 17:33:29
1 2020-04-13 17:35:16
2 2020-04-13 17:36:57
3 2020-04-13 17:38:36
4 2020-04-13 17:40:15
5 2020-04-13 17:41:54
6 2020-04-13 17:43:32
7 2020-04-13 17:45:10
8 2020-04-13 17:46:48
9 2020-04-13 17:48:25
10 2020-04-13 17:50:01
11 2020-04-13 17:51:37
12 2020-04-13 17:53:12
13 2020-04-13 17:54:48
14 2020-04-13 17:56:23
15 2020-04-13 17:57:57
16 2020-04-13 17:59:30
17 2020-04-13 18:01:04
18 2020-04-13 18:02:38
19 2020-04-13 18:04:12
During-farm period
20 2020-04-13 18:05:45
21 2020-04-13 18:10:08
22 2020-04-13 18:12:27
23 2020-04-13 18:15:22
24 2020-04-13 18:17:42
25 2020-04-13 18:19:53
26 2020-04-13 18:22:02
27 2020-04-13 18:25:16
28 2020-04-13 18:28:04
29 2020-04-13 18:30:53
30 2020-04-13 18:33:43
31 2020-04-13 18:35:55
32 2020-04-13 18:39:13
33 2020-04-13 18:42:02
34 2020-04-13 18:43:41
35 2020-04-13 18:45:48
36 2020-04-13 18:48:23
37 2020-04-13 18:50:27
38 2020-04-13 18:53:38
39 2020-04-13 18:55:47
Post-farm period
40 2020-04-13 18:59:32
41 2020-04-13 19:00:54
42 2020-04-13 19:02:16
43 2020-04-13 19:03:38
44 2020-04-13 19:05:00
45 2020-04-13 19:06:23
46 2020-04-13 19:07:46
47 2020-04-13 19:09:09
48 2020-04-13 19:10:31
49 2020-04-13 19:11:54
50 2020-04-13 19:13:16
51 2020-04-13 19:14:39
52 2020-04-13 19:16:00
53 2020-04-13 19:17:22
54 2020-04-13 19:18:45
55 2020-04-13 19:20:07
56 2020-04-13 19:21:29
57 2020-04-13 19:22:50
58 2020-04-13 19:24:11
59 2020-04-13 19:25:32
Reading in parameter values
batch_testgnu01_srep_5
Initializing response variable storage objecs
Rep 0 2020-04-13 17:33:28
Initializing starting population
Pre-farm period
0 2020-04-13 17:33:29
1 2020-04-13 17:35:16
2 2020-04-13 17:36:58
3 2020-04-13 17:38:38
4 2020-04-13 17:40:16
5 2020-04-13 17:41:55
6 2020-04-13 17:43:32
7 2020-04-13 17:45:09
8 2020-04-13 17:46:46
9 2020-04-13 17:48:23
10 2020-04-13 17:49:59
11 2020-04-13 17:51:36
12 2020-04-13 17:53:12
13 2020-04-13 17:54:47
14 2020-04-13 17:56:22
15 2020-04-13 17:57:56
16 2020-04-13 17:59:30
17 2020-04-13 18:01:05
18 2020-04-13 18:02:39
19 2020-04-13 18:04:13
During-farm period
20 2020-04-13 18:05:46
21 2020-04-13 18:09:35
22 2020-04-13 18:11:51
23 2020-04-13 18:15:21
24 2020-04-13 18:18:19
25 2020-04-13 18:20:37
26 2020-04-13 18:22:17
27 2020-04-13 18:25:32
28 2020-04-13 18:27:42
29 2020-04-13 18:30:27
30 2020-04-13 18:32:04
31 2020-04-13 18:35:17
32 2020-04-13 18:38:39
33 2020-04-13 18:40:19
34 2020-04-13 18:42:28
35 2020-04-13 18:45:37
36 2020-04-13 18:47:10
37 2020-04-13 18:49:47
38 2020-04-13 18:53:37
39 2020-04-13 18:56:23
Post-farm period
40 2020-04-13 18:59:12
41 2020-04-13 19:00:34
42 2020-04-13 19:01:56
43 2020-04-13 19:03:18
44 2020-04-13 19:04:41
45 2020-04-13 19:06:03
46 2020-04-13 19:07:26
47 2020-04-13 19:08:50
48 2020-04-13 19:10:13
49 2020-04-13 19:11:37
50 2020-04-13 19:13:01
51 2020-04-13 19:14:25
52 2020-04-13 19:15:48
53 2020-04-13 19:17:11
54 2020-04-13 19:18:35
55 2020-04-13 19:19:59
56 2020-04-13 19:21:23
57 2020-04-13 19:22:48
58 2020-04-13 19:24:12
59 2020-04-13 19:25:36
```

**Comment by nclowell** [comment]
I just had my first unsuccessful job using GNU Parallel (not sure what happened) and I just wanted to mention that nothing got written to slurm out (which is why I don't know what happened) even though the job ran for 18 hours and I would have expected a lot of stuff in there similar to the previous post in this thread. Just wanted to share this in case anyone else in the group tries GNU Parallel!
